src/test/integration/base.py

```diff
@@ -0,0 +1,141 @@
+SERVER_SOCKET_PATH = TESTING_DIRECTORY + "server.socket"
+WORKER_SOCKET_PATH = TESTING_DIRECTORY + "worker-%s.socket"
+SERVER_REMOTE_MODULE = "test.integration.remote.server"
+WORKER_REMOTE_MODULE = "test.integration.remote.%s"
+if getuser() == "travis":
+    COMMAND_STRING = "~/virtualenv/python3.7/bin/python3 -m %s"
+else:
+    COMMAND_STRING = "python3 -m %s"
+
+
+class TestWorkerProxy(WorkerProxyBase):
+
+    def _get_ssh_connection(self, ssh_config: SSHConfig) -> ISSHConnection:
+        return SSHConnection(
+            ssh_config=ssh_config,
+            remote_module=WORKER_REMOTE_MODULE % self._uid,
+            command_string=COMMAND_STRING
+        )
+
+
+class TestWorkerProxyFactory(WorkerProxyFactoryBase):
+
+    def _create_proxy(self, work_machine: WorkMachine) -> IWorkerProxy:
+        return TestWorkerProxy(uid=work_machine.uid, ssh_config=None)  # FIXME need to get worker ssh config
+
+
+class TestJobCenter(JobCenter):
+
+    @staticmethod
+    def _read_config() -> ServerConfig:
+        with open(SERVER_CONF_PATH) as f:
+            return ServerConfig.from_string(f.read())
+
+    def _get_proxy_factory(self) -> WorkerProxyFactoryBase:
+        return TestWorkerProxyFactory(self._database)
+
+
+class TestJobWorker(JobWorker):
+    def __init__(self, index: int):
+        self._index = index
+        super().__init__(
+            config_path=WORKER_CONF_PATH % self._index, socket_path=WORKER_SOCKET_PATH,
+            remote_module=SERVER_REMOTE_MODULE, command_string=COMMAND_STRING)
+
+
+class IntegrationTest(TestCase):
+
+    def setUp(self) -> None:
+        Path(TESTING_DIRECTORY).mkdir(parents=True, exist_ok=True)
+        with open(SERVER_CONF_PATH, "w") as f:
+            f.write(str(self.server_config))
+        self._server = TestJobCenter(socket_path=SERVER_SOCKET_PATH, database_name=DATABASE_NAME)
+        Thread(target=self._server.run, name="server-main", daemon=True).start()
+
+        with open(SSH_CONF_PATH, "w") as f:
+            f.write(str(self.ssh_config))
+
+        self._clients: List[UserClientCLIHandler] = []
+        for i in range(self.num_clients):
+            client = UserClientCLIHandler(config_path=SSH_CONF_PATH)
+            self._clients.append(client)
+
+        self._workers: List[JobWorker] = []
+        for i in range(self.num_workers):
+            with open(WORKER_CONF_PATH % i, "w") as f:
+                f.write(str(self.get_worker_config(i)))
+            worker = TestJobWorker(index=i)
+            Thread(target=worker.run, name="worker%s" % i, daemon=True).start()
+            self._workers.append(worker)
+        sleep(1)
+
+    def tearDown(self) -> None:
+        database = self._server._database
+        for table in reversed(database._metadata.sorted_tables):
+            database.engine.execute(table.delete())
+
+    @property
+    def ssh_config(self) -> SSHConfig:
+        return SSHConfig(hostname="localhost", username=getuser())
```

Comment:
using 127.0.0.1 is more versatile, I had 127.0.0.1 working on my laptop but localhost not. It is all a matter of configuration in the end, but 127.0.0.1 should work in more cases.